implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs

```diff
@@ -67,10 +65,12 @@ impl TcpSendWorker {
         stream: Option<TcpStream>,
         peer: SocketAddr,
         internal_addr: Address,
-        heartbeat: DelayedEvent<TcpSendWorkerMsg>,
     ) -> Self {
+        let keepalive = TcpKeepalive::new().with_time(Duration::from_secs(4));
         let (rx, tx) = match stream {
```

Comment:
As far as I remember, there are 2 cases here: if we've been running a listener and just accepted new connection - in this case stream here is Some. In case we're initiating connection, stream is created in TcpSendWorker::initialize. According to this recently created issue #3480 we wanted to run keepalive only if we're initiating connection, but that was true for our custom keep-alive mechanism, let me clarify if that's also what we want with tcp-native keep-alive. I'll get back to you

Reply:
Yes I can do that. I might have some questions regarding the setup for the test along the way but I will see what I can do. 

I can add a flag that this is only tested when explicitly included.

Reply:
@SanjoDeundiak I'm not really sure what you mean by cloud node?

Looking at the code right now, I have some questions regarding the test.
Would it make sense to use the setup of the `send_receive` function in  `ockam_transport_tcp/tests/send_receive.rs`?:

![image](https://user-images.githubusercontent.com/57916112/191336117-9a946a75-50ac-42bf-a203-b52271c37441.png)
Although I am really not sure that I understand how the Sender is setup. 
Does this line basically mean that we are creating a tcp sender: 
`let r = route![(TCP, listener_address.to_string()), "echoer"];` ?

So fist we create a `TcpTransport` and then call listen() which returns a `SocketAddr` which is then passed in a tuple alongside `TCP` to `route!`, which does what exactly with the two values in the tuple? Is `TCP` not just a kind of `TransportType`?

The  result of all this is that we have a route of `1#127.0.0.1:8085 => 0#echoer` meaning we are sending a message to a `TcpSendWorker` which then sends the message to an echoer worker (which is not tcp)? 

Basically we want the `TcpSendWorker` to send keepalive packets. Does this mean i have to setup another `TcpSendWorker` because I think I have not fully understood how exactly the `send_receive` test works exactly and if this is really the setup where the keepalive packets should actually be exchanged.

Sorry for the many questions and if they are not particularly well formulated.

Reply:
Hi @SanjoDeundiak thanks for your help. It took me a second but I finally understood, what needs to be tested. I still wasn't sure what you meant by the cloud node and if that was something that I needed to setup in the code. But after some investigating I found out that `1.node.ockam.network:4000` is actually a deployed ockam node on an ec2 instance (?) that has an echo worker running on it and that is available to connect to. Not sure why I didn't get that earlier.
Anyways I have added a test that connects to it, waits for 10 minutes (or even less)  and then tries to send another message. Unfortunately it is not working as expected because the connection is lost after some time and hence I assume the tcp_keepalive is not working properly:
![Screenshot from 2022-10-03 00-39-31](https://user-images.githubusercontent.com/57916112/193633914-7416a45d-593e-4b3a-83a8-695a8dcd9ec8.png)
While the "old" version with the heartbeats worked and the connection is not closed:
![Screenshot from 2022-10-03 00-29-23](https://user-images.githubusercontent.com/57916112/193634156-8286943f-e451-4f92-8153-cceba65c52d0.png)

I tried adding further options to the the tcp_keepalive option to prevent the connection from being dropped, but nothing has worked thus far: 
![image](https://user-images.githubusercontent.com/57916112/193634604-231dc6c7-1743-4841-8568-9b750dcf7031.png)
(btw. as you can see in the test I had to set the timeout feature of the test to some big number because otherwise I always got the custom error message that the test timed out).

I will have to further investigate this and play around with the options a bit more. Maybe there is also something that I am not understanding correctly about TCP sockets and keepalive that prevents this from working.

Reply:
Hi @SanjoDeundiak, sorry this is taking so long. So i realized that my test was using some function that ends up calling `TcpSendWorker::new_pair()` without passing it `Some(Stream)`. That is only done by `TcpListenProcessor.process()` (and then I can see that tcp keepalive is indeed enabled for the port).

But if I understood you correctly that is not what we want, i.e. we don't want the keepalive option to be enabled when we are listening, but only if we are initiating the connection. But the only place where this gets called is through `TcpRouter.handle_connect()` which calls `TcpSendWorker::start_pair` but passes `None` to the stream parameter.

That means there needs to be some stream created for the connection initiation side as well and not just for the listener side as it is in `TcpListenProcessor.process` in listener.rs:
```rust
async fn process(&mut self, ctx: &mut Self::Context) -> Result<bool> {
        debug!("Waiting for incoming TCP connection...");

        // Wait for an incoming connection
        let (stream, peer) = self.inner.accept().await.map_err(TransportError::from)?;
        debug!("TCP connection accepted");

        let handle_clone = self.router_handle.async_try_clone().await?;
        // And create a connection worker for it
        let (worker, pair) =
            TcpSendWorker::new_pair(handle_clone, Some(stream), peer, Vec::new()).await?;
```
We can see here that stream is created by calling accept on the `TcpListener`. Maybe it's possible to call connect  on the `TcpRouterHandle`  or something like that and have it return a stream as well which could be used when creating a new `TcpSendWorker` not through a listener. 

Does that make any sense to you? Or do you have any suggestions for this?

Reply:
Hi @SanjoDeundiak or @mrinalwadhwa is there maybe any comment on this. I think I can resolved this issue, but I am currently stuck at the points mentioned above and I am not really sure how to proceed. 

Thanks a lot for any guidance or comment

Reply:
@phillyphil91 sorry, I was have been unavailable for some time! For connecting side TCP stream is created inside `TcpSendWorker::initialize` function, which is called by our runtime when that worker's run loop is about to start

Reply:
hey @SanjoDeundiak thanks for the update and help. I think that was the missing piece to make it work. Not sure why i didn't see earlier that the keepalive option on the socket needed to be created on `TcpSendWorker::initialize`. 

Everything works fine now, I believe. When the tcp keepalive option is enabled for the tcp socket, the connection to ockam cloud node does not die as it would without the option. I also verified that keepailve is enabled for the connection via netsat (it is not enabled without the additional code)

A few things for clarification around configuring the tcp keepalive option and the test. Feel free to confirm or suggest changes:

TCP:
- The tcp keepalive probes will be send out after 5 minutes
- There will be 60 probes send out, each after 5 seconds adding a total of 300 seconds, i.e. keeping the connection alive for another 5 minutes

Test:
- After having send a message to the ockam cloud node the thread sleeps for 350 seconds and the tries to send another message
- Tests timeout after 30 seconds so I added a timeout time of 400 seconds so the test will run completely
- I added a `#[ignore]` flag to the test to require explicitly running that test only if necessary to not hold up the CI pipeline

Let me know if that works for you or if there is anything else that needs to be changed. 

Maybe one more question: When sending the first message to the ockam node we create a tcp connection to it from our node and that is the same one that is being used after the threads sleeps, right? I mean otherwise the tcp keepalive code would not run the way it does and there wouldn't be the info message if tcp keepalive is not enabled:
![image](https://user-images.githubusercontent.com/57916112/198830300-7a8f0834-dbdb-41df-8d41-f49d3bc885a4.png)
But I just wanted to confirm that assumption.

Reply:
@SanjoDeundiak I just realized that since the tcp keepalive is enabled in the `TcpSendWorker::initialize` only, the Heartbeat implementation that was there before can remain at the other places in the code, right? Should i add it to the other code, e.g.`TcpSendWorker::new` etc. again where it was before? Or do you not want the Heartbeat anywhere anymore?

Reply:
@phillyphil91 I think we don't need Heartbeats in TCP anymore.